### PR TITLE
#6 - WIP: Added basic readme

### LINKED ISSUE #6: Documentation: Write readme
Primary contents:
- What the project is about
- Security and context handling
- Contributing
- Testing
- How to run
- Release strategy

### PR DESCRIPTION
**Pull request by TStrecha**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+
+Each new logic should be tested manually through options mentioned above, but also tested using automatic tests - either unit or integration tests to ensure that each and every time something changes, pipeline tests all features and in case something fails, we'll know about it. Automatic tests should test most edge cases possible and ensure that the code works properly using these tests.
+
+<a name="branch-standards"></a>
+### Branch standards
+Main branch is a protected branch that has the latest full release on. Develop branch on the other hand should have features until they are fully released and pushed into main branch.
+
+Every change to code should be in a different branch and should only be merged into develop branch through pull request. Pull requests are always onto develop branch, excluding release pull requests. Pull request has to be reviewed first and accepted before merging.
+
+Naming strategies for branches are:
```
Comment (TStrecha) [inline README.md]:
```suggestion
Naming strategy for branches is:
```

File: README.md [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+the duration of this token and the secret key. Secret keys are different on each configuration profile to avoid tokens being easily mock-able.
+
+Tokens hold user context, which holds detailed information about currently logged user, it's permissions and which account is user logged as.
+
+Users are divided into 2 roles - users and admins. Users are regular users and can do limited operations in contexts according to their permissions, whereas admins can log as anyone and do anything to every account.
+They also have some specific endpoints to handle, debug and fix some issues among the regular users on production.
+
+<a name="security-filter-chain"></a>
+### Security filter chain
+Spring filter will  decline every request, apart from open api, swagger ui, spring error pages and basic authorization - being registration and login.
```
Comment (TStrecha) [inline README.md]:
```suggestion
Spring filter will decline every request, apart from open api, swagger ui, spring error pages and basic authorization - being registration and login.
```

File: README.md [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+
+<a name="user-context"></a>
+### User context
+User context defines everything that is needed about currently logged user, including his permissions. 
+Users can log as different user, this is defined by `UserContext.loggedAs` and when user changes context, new token has to be generated for him.
+More about this feature in the [Product description](#product-description) section.
+
+Permissions in the context define permissions which has current logged user to the user which he is currently logged as.
+
+Spring boot security context holds user context and can be retrieved at any point. For a case, where context is needed in the controller endpoint,
```
Comment (TStrecha) [inline README.md]:
```suggestion
Spring boot security context holds user context and can be retrieved at any point. If context is needed in an endpoint,
```

File: README.md [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+parameter resolution logic is provided and can be forced by custom annotation `@InjectUserContext`.
```
Comment (TStrecha) [inline README.md]:
```suggestion
custom annotation `@InjectUserContext` can be used.
```

File: README.md [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+required permissions defined in the value of this annotation.
+
+<a name="how-to-run"></a>
+## How to run
+For running the application locally, first clone the repository using
+```shell
+git clone https://github.com/TStrecha/time-tracker.git
+```
+
+Local [postgres server](https://www.postgresql.org/) is also needed with user `timetracker` and password `timetracker`. By default, host of postgres server for local profile is set to `localhost:5432`, change it, if required, but machine specific changes shouldn't be pushed onto git.
```
Comment (TStrecha) [inline README.md]:
```suggestion
Run a local [postgres server](https://www.postgresql.org/) is with an user `timetracker` having a password of `timetracker`. By default, host of the postgres server for set to `localhost:5432` in local profile. Change it, if required, but machine specific changes shouldn't be pushed onto git.
```

File: README.md [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+- [Contributing](#contributing)
+  - [Code standards](#code-standards)
+  - [Changing an entity](#changing-an-entity)
+  - [Testing](#testing)
+  - [Branch standards](#branch-standards)
+  - [Development tracking](#development-tracking)
+
+<a name="product-description"></a> 
+## Product description
+_To be defined._
```
Comment (TStrecha) [inline README.md]: TODO

File: README.md [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+
+<a name="development-strategy"></a>
+## Development strategy
+The development strategy defines how the development is done for both backend and frontend applications and how the product is released to the end user. At first the focus should be making the minimum viable product, so that the foundation of the whole product is set and can be build upon.
+
+The step of develop is for this reason divided into releases - these releases define what is the end goal and what the product should look like after this release is done. Release should be defined before it's beginning and shouldn't be changed during the progression.
+
+<a name="release-1"></a>
+### Release 1
+
```
Comment (TStrecha) [inline README.md:104]: TODO

File: README.md [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+Or when testing the application, for unit tests run:
+```shell
+mvn test -Dtest=!*IT
+```
+And for integration tests run:
+```shell
+mvn test -Dtest=*IT
+```
+Local test profile is set automatically through abstract integration test class. No need to worry about that.
```
Comment (TStrecha) [inline README.md]:
```suggestion
Local test profile is set automatically through abstract integration test class using @BeforeEach. No need to worry about that.
```

File: README.md [hunk 9]
```diff
@@ -0,0 +1,171 @@
...
+The development strategy defines how the development is done for both backend and frontend applications and how the product is released to the end user. At first the focus should be making the minimum viable product, so that the foundation of the whole product is set and can be build upon.
+
+The step of develop is for this reason divided into releases - these releases define what is the end goal and what the product should look like after this release is done. Release should be defined before it's beginning and shouldn't be changed during the progression.
+
+<a name="release-1"></a>
+### Release 1
+
+<a name="release-2"></a>
+### Release 2
+_Not yet defined._
```
Comment (TStrecha) [inline README.md]: todo?

File: README.md [hunk 10]
```diff
@@ -0,0 +1,171 @@
...
+
+<a name="changing-entity"></a>
+### Changing an entity
+Since the application uses migration, specifically liquibase, when changing an entity, it's required to also write a liquibase script. That can be done either by creating script manually, or running
+```shell
+mvn clean install -DskipTests liquibase:generateChangeLog
+```
+
+this will automatically generate diff in the `src/main/resources/liquibase` folder. This file has to be checked before getting onto git.
+New generated change log's name should start with 4 digit long unique string match this regex
```
Comment (TStrecha) [inline README.md]:
```suggestion
New generated change log's name should start with 4 digit long unique string match this regex:
```

File: README.md [hunk 11]
```diff
@@ -0,0 +1,171 @@
...
+Also has to contain author's name and id of each change set has to match the name of the change log file with some unique number suffix, in regex words:
+```regexp
+\d{4}-([a-zA-Z-])+-\d+
+```
+
+<a name="testing"></a>
+### Testing
+Testing endpoints is possible through IntelliJ Idea http client, although it is not a recommended option. Better way to test endpoints is using [Postman](https://www.postman.com/)
+
+Each new logic should be tested manually through options mentioned above, but also tested using automatic tests - either unit or integration tests to ensure that each and every time something changes, pipeline tests all features and in case something fails, we'll know about it. Automatic tests should test most edge cases possible and ensure that the code works properly using these tests.
```
Comment (TStrecha) [inline README.md]:
```suggestion
Each new logic should be tested manually through options mentioned above, but also tested using automatic tests - either unit or integration tests to ensure that each and every time something changes, pipeline tests all features and in case something fails, we'll know about it. Automatic tests should test the most edge cases possible and ensure that the code works properly using these tests.
```

File: README.md [hunk 12]
```diff
@@ -0,0 +1,171 @@
...
+
+<a name="branch-standards"></a>
+### Branch standards
+Main branch is a protected branch that has the latest full release on. Develop branch on the other hand should have features until they are fully released and pushed into main branch.
```
Comment (TStrecha) [inline README.md]:
```suggestion
Main branch is a protected branch that has the latest full release on. Develop branch on the other hand should have features until they are fully released and merged into the main branch.
```

File: README.md [hunk 13]
```diff
@@ -0,0 +1,171 @@
...
+
+Every change to code should be in a different branch and should only be merged into develop branch through pull request. Pull requests are always onto develop branch, excluding release pull requests. Pull request has to be reviewed first and accepted before merging.
```
Comment (TStrecha) [inline README.md]:
```suggestion
Every change to the code should be in a different branch and should only be merged into develop branch through pull request. Pull requests are always onto develop branch, excluding release pull requests. Pull request has to be reviewed first and accepted before merging.
```

File: README.md [hunk 14]
```diff
@@ -0,0 +1,171 @@
...
+```text
+[feature|bug]/[issue number]-[simple change description]
+```
+
+For example:
+```text
+feature/6-added-readme
+```
+
+Make sure to mark pull request as draft, if the changes are not complete. Also naming of the pull request should follow naming strategy of branches.
```
Comment (TStrecha) [inline README.md]:
```suggestion
Make sure to mark a pull request as a draft, if the changes are not complete. Also naming of the pull request should follow naming strategy of branches.
```

File: README.md [hunk 15]
```diff
@@ -0,0 +1,171 @@
...
+
+Commits should follow simple rule:
+```text
+#[issue id] - [Commit message]
+```
+
+<a name="development-tracking"></a>
+### Development tracking
+At no point should anyone be working at change that has no issue behind it. Every change should be part of an issue and every pull request should be linked to this issue. If no issue is provided, create one instead.
```
Comment (TStrecha) [inline README.md]:
```suggestion
At no point should anyone be working at a change that has no issue behind it. Every change should be part of an issue and every pull request should be linked to this issue. If no issue is provided, create one instead.
```

File: README.md [hunk 16]
```diff
@@ -0,0 +1,171 @@
...
+  - [Development tracking](#development-tracking)
+
+<a name="product-description"></a> 
+## Product description
+_To be defined._ 
+
+<a name="technical-description"></a>
+## Technical description
+Time Tracker runs on Spring boot version 3.0 and requires Java 17. 
+Access to api and api generation, including frontend client, is done by open api.
```
Comment (TStrecha) [inline README.md]:
```suggestion
Access to endpoints and api generation, including frontend client, is done by open api.
```

File: README.md [hunk 17]
```diff
@@ -0,0 +1,171 @@
...
+
+Local database details are explained in the [How to run](#how-to-run) section below.
+
+For object mapping, mapstruct is used with component model parameter being in the pom with the addition of jackson ObjectMapper.
+
+<a name="security"></a>
+### Security
+Security is provided through spring security context. User can log in using the
+`/auth/login` endpoint using their information provided in the registration.
+Once user is successfully authenticated, JWT access token is created for that user. Application settings set properties of the JWT token -
```
Comment (TStrecha) [inline README.md]:
```suggestion
Once user is successfully authenticated, a JWT access token is created for that user. Application settings set properties of the JWT token -
```

File: README.md [hunk 18]
```diff
@@ -0,0 +1,171 @@
...
+the duration of this token and the secret key. Secret keys are different on each configuration profile to avoid tokens being easily mock-able.
+
+Tokens hold user context, which holds detailed information about currently logged user, it's permissions and which account is user logged as.
```
Comment (TStrecha) [inline README.md]:
```suggestion
Tokens hold an user context, which holds detailed information about currently logged user, it's permissions and which account is user logged as.
```

File: README.md [hunk 19]
```diff
@@ -0,0 +1,171 @@
...
+For example:
+```text
+feature/6-added-readme
+```
+
+Make sure to mark pull request as draft, if the changes are not complete. Also naming of the pull request should follow naming strategy of branches.
+
+Commits should follow simple rule:
+```text
+#[issue id] - [Commit message]
```
Comment (TStrecha) [inline README.md]:
```suggestion
#[issue id] - [commit message]
```
